### PR TITLE
OAK-11073 : added conversion utility methods to convert iterable/iter…

### PR DESCRIPTION
**Pull request by rishabhdaim**
…ator to set/list

### CODE REVIEW

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ * Utility methods for collections conversions.
+ */
+public class CollectionUtils {
+
+    private CollectionUtils() {
+        // no instances for you
+    }
+
+    /**
+     * Convert an iterable to a list.The retuning list is mutable and supports all optional operations.
```
Comment (reschke) [inline oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]: "Convert an iterable to a list. The returning list is mutable and supports all optional operations."

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+     * @param iterable the iterable to convert
+     * @return the list
+     * @param <T> the type of the elements
+     */
+    public static <T> List<T> toList(final Iterable<T> iterable) {
+        return IterableUtils.toList(iterable);
+    }
+
+    /**
+     * Convert an iterator to a list. The retuning list is mutable and supports all optional operations.
```
Comment (reschke) [inline oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]: "returning"

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+     * @param iterator the iterator to convert
+     * @return the list
+     * @param <T> the type of the elements
+     */
+    public static <T> List<T> toList(final Iterator<T> iterator) {
+        return IteratorUtils.toList(iterator);
+    }
+
+    /**
+     * Convert an iterable to a set. The retuning set is mutable and supports all optional operations.
```
Comment (reschke) [inline oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]: (same)

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+     * @param iterable the iterable to convert
+     * @return the set
+     * @param <T> the type of the elements
+     */
+    public static <T> Set<T> toSet(final Iterable<T> iterable) {
+        return StreamSupport.stream(iterable.spliterator(), false).collect(Collectors.toSet());
+    }
+
+    /**
+     * Convert an iterator to a set. The retuning set is mutable and supports all optional operations.
```
Comment (reschke) [inline oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]: (dito)

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/package-info.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+/**
+ * Annotations for internal use in Oak.
+ */
```
Comment (reschke) [inline oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/package-info.java:23]: don't we need a vesion number here as well?

**Review by reschke — COMMENTED** [review]
Mostly ok except for typos and one question re package version.

Before we go ahead with this, I'd like to see @thomasmueller's opinion though.

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/package-info.java (hunk 5 above)
Comment (rishabhdaim) [reply oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/package-info.java:23]: added in https://github.com/apache/jackrabbit-oak/pull/1679/commits/35928c9399aec11d7aaf8e6fc6cfcce4b43e902f

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java (hunk 1 above)
Comment (rishabhdaim) [reply oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]: addressed in https://github.com/apache/jackrabbit-oak/pull/1679/commits/35928c9399aec11d7aaf8e6fc6cfcce4b43e902f

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java (hunk 2 above)
Comment (rishabhdaim) [reply oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]: addressed in https://github.com/apache/jackrabbit-oak/pull/1679/commits/35928c9399aec11d7aaf8e6fc6cfcce4b43e902f

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java (hunk 3 above)
Comment (rishabhdaim) [reply oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]: addressed in https://github.com/apache/jackrabbit-oak/pull/1679/commits/35928c9399aec11d7aaf8e6fc6cfcce4b43e902f

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java (hunk 4 above)
Comment (rishabhdaim) [reply oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]: addressed in https://github.com/apache/jackrabbit-oak/pull/1679/commits/35928c9399aec11d7aaf8e6fc6cfcce4b43e902f

**Comment by reschke** [comment]
We'll also need to add the package to the export list in the POM. Otherwise this won't be available in OSGi.

File: oak-commons/pom.xml [hunk 6]
```diff
@@ -53,7 +53,8 @@
               org.apache.jackrabbit.oak.commons.json,
               org.apache.jackrabbit.oak.commons.log,
               org.apache.jackrabbit.oak.commons.sort,
-              org.apache.jackrabbit.oak.commons.properties
+              org.apache.jackrabbit.oak.commons.properties,
```
Comment (reschke) [inline oak-commons/pom.xml]: keep it sorted :-)

**Review by mbaedke — APPROVED** [review]

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    /**
+     * Convert an iterable to a list.The returning list is mutable and supports all optional operations.
+     * @param iterable the iterable to convert
+     * @return the list
+     * @param <T> the type of the elements
+     */
+    public static <T> List<T> toList(final Iterable<T> iterable) {
+        return IterableUtils.toList(iterable);
```
Comment (thomasmueller) [inline oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]:
```suggestion
        List<String> result = new ArrayList<String>();
        iterable.forEach(result::add);
        return result;
```
    iterable.forEach(result::add);

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    /**
+     * Convert an iterator to a list. The returning list is mutable and supports all optional operations.
+     * @param iterator the iterator to convert
+     * @return the list
+     * @param <T> the type of the elements
+     */
+    public static <T> List<T> toList(final Iterator<T> iterator) {
+        return IteratorUtils.toList(iterator);
```
Comment (thomasmueller) [inline oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]:
```suggestion
        List<String> result = new ArrayList<String>();
        iterator.forEachRemaining(result::add);
        return result;
```

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.commons.collections;
+
+import org.apache.commons.collections4.IterableUtils;
+import org.apache.commons.collections4.IteratorUtils;
```
Comment (thomasmueller) [inline oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]: These dependencies are not needed (see below).

**Review by thomasmueller — CHANGES_REQUESTED** [review]
There a no unit tests.

The new dependencies are not needed.

**Comment by reschke** [comment]
FWIW: the dependency isn't new.

File: oak-commons/pom.xml [hunk 10]
```diff
@@ -103,6 +104,11 @@
       <artifactId>oak-shaded-guava</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>org.apache.commons</groupId>
+      <artifactId>commons-collections4</artifactId>
+      <version>4.4</version>
```
Comment (reschke) [inline oak-commons/pom.xml]: version should be removed, it's in the parent pom

**Review by reschke — CHANGES_REQUESTED** [review]

**Review by thomasmueller — APPROVED** [review]

File: oak-commons/pom.xml (hunk 10 above)
Comment (rishabhdaim) [reply oak-commons/pom.xml]: addressed in https://github.com/apache/jackrabbit-oak/pull/1679/commits/d44677ee9468e3dc173ab4fd56c7a6a7445aba00

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+ * Utility methods for collections conversions.
+ */
+public class CollectionUtils {
+
+    private CollectionUtils() {
+        // no instances for you
+    }
+
+    /**
+     * Convert an iterable to a list.The returning list is mutable and supports all optional operations.
```
Comment (reschke) [inline oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java]:
```suggestion
     * Convert an iterable to a list. The returning list is mutable and supports all optional operations.
```

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        return result;
+    }
+
+    /**
+     * Convert an iterable to a set. The returning set is mutable and supports all optional operations.
+     * @param iterable the iterable to convert
+     * @return the set
+     * @param <T> the type of the elements
+     */
+    public static <T> Set<T> toSet(final Iterable<T> iterable) {
```
Comment (reschke) [inline oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java:66]: For consistency, could we implement this in an analogous way (forEachRemaining)?

File: oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtilsTest.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Test
+    public void iteratorToSet() {
+        // create an iterable
+        final Set<String> data = Set.of("one", "two", "three");
+        final Iterable<String> iterable = new SimpleIterable<>(data);
+
+        Assert.assertEquals(data, CollectionUtils.toSet(iterable.iterator()));
+    }
+}
```
Comment (reschke) [inline oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtilsTest.java:67]: Maybe add null values to tests?

**Review by reschke — CHANGES_REQUESTED** [review]
Just a typo and a few minor questions/suugestions.

File: oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtilsTest.java (hunk 13 above)
Comment (rishabhdaim) [reply oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtilsTest.java:67]: added in https://github.com/apache/jackrabbit-oak/pull/1679/commits/cf9b2bc93de33e61604d0559fd000c468d46c9f2

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java (hunk 12 above)
Comment (rishabhdaim) [reply oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/collections/CollectionUtils.java:66]: addressed in https://github.com/apache/jackrabbit-oak/pull/1679/commits/cf9b2bc93de33e61604d0559fd000c468d46c9f2

**Review by reschke — APPROVED** [review]

**Pull request merged**